### ISSUE TITLE
BUG: unable to handle page fault for address: fffffffffffffffd

### ISSUE THREAD

**Issue by marcin-github** [koverstreet/bcachefs#76]
Kernel on hypervisor hardly crashed (memory corruption, probably related to nfs/rpc). I had to do hard reset bare metal hypervisor. After this bcahefs inside VM got corrupted.
kernel is on  commit 66925f9 inside vm. Access to filesystem ends with `BUG: unable to...`

```
[   20.490192] tun: Universal TUN/TAP device driver, 1.6
[   20.491564] systemd-udevd[3311]: link_config: autonegotiation is unset or enabled, the speed and duplex are not writable.
[  103.083399] BUG: unable to handle page fault for address: fffffffffffffffd
[  103.083779] #PF: supervisor read access in kernel mode
[  103.084081] #PF: error_code(0x0000) - not-present page
[  103.084412] PGD 5e0c067 P4D 5e0c067 PUD 5e0e067 PMD 0
[  103.084722] Oops: 0000 [#2] SMP PTI
[  103.084978] CPU: 3 PID: 3451 Comm: rrdcached Tainted: G      D    OE   T 5.2.0-73630-g66925f93919a #2
[  103.085499] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.12.1-0-ga5cab58e9a3f-prebuilt.qemu.org 04/01/2014
[  103.086086] RIP: 0010:bch2_inode_find_by_inum_trans+0x88/0xb0 [bcachefs]
[  103.086467] Code: ff ff 48 89 c3 41 89 c5 76 12 48 8d 65 d8 44 89 e8 5b 41 5a 41 5c 41 5d 41 5e 5d c3 48 89 c7 41 bd fe ff ff ff e8 38 c4 fc ff <80> 78 02 08 48 89 d6 74 0d 48 89 de 4c 89 e
7 e8 f4 c9 fc ff eb ca
[  103.087380] RSP: 0018:ffffafaa41ed3848 EFLAGS: 00010246
[  103.087709] RAX: fffffffffffffffb RBX: ffffafaa41ed3928 RCX: 0000000000000001
[  103.088073] RDX: 0000000000000000 RSI: 0000000000000001 RDI: ffff9ac97f1a0000
[  103.088436] RBP: ffffafaa41ed3888 R08: 0000000000000000 R09: 0000000000002f23
[  103.088824] R10: 0000000000000080 R11: ffffafaa41ed3928 R12: ffffafaa41ed3898
[  103.089205] R13: 00000000fffffffe R14: ffffafaa41ed3b18 R15: 0000000000000000
[  103.089613] FS:  00007fba311dd900(0000) GS:ffff9ac67fcc0000(0000) knlGS:0000000000000000
[  103.090088] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  103.090408] CR2: fffffffffffffffd CR3: 00000001fbb1e001 CR4: 0000000000360ea0
[  103.090792] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  103.091177] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  103.091545] Call Trace:
[  103.091801]  bch2_inode_find_by_inum+0x6c/0xd0 [bcachefs]
[  103.092124]  ? bch2_inode_find_by_inum+0x4a/0xd0 [bcachefs]
[  103.092447]  bch2_vfs_inode_get+0x88/0xd0 [bcachefs]
[  103.092772]  bch2_lookup+0x4a/0x60 [bcachefs]
[  103.093070]  __lookup_slow+0xa3/0x160
[  103.093318]  lookup_slow+0x38/0x50
[  103.093597]  walk_component+0x218/0x3d0
[  103.093863]  path_lookupat.isra.56+0x76/0x260
[  103.094130]  filename_lookup+0xc2/0x150
[  103.094382]  vfs_statx+0x85/0xf0
[  103.094612]  __do_sys_newstat+0x47/0x80
[  103.094953]  do_syscall_64+0x4f/0xf0
[  103.107282]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  103.107619] RIP: 0033:0x7fba3234ea89
[  103.107865] Code: 00 48 83 ec 18 64 48 8b 0c 25 28 00 00 00 48 89 4c 24 08 31 c9 83 ff 01 77 4f 48 89 f0 48 89 c7 48 89 d6 b8 04 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 1f 48 8b 4c 24 08 64 48 33 0c 25 28 00 00 00
[  103.108730] RSP: 002b:00007ffd7c30dc00 EFLAGS: 00000246 ORIG_RAX: 0000000000000004
[  103.109201] RAX: ffffffffffffffda RBX: 0000000000000000 RCX: 00007fba3234ea89
[  103.109548] RDX: 00007ffd7c30dc80 RSI: 00007ffd7c30dc80 RDI: 000055fa23d8fb20
[  103.109937] RBP: 00007ffd7c30dc50 R08: 0000000000000000 R09: 00007ffd7c30ee67
[  103.110281] R10: 000000000000005b R11: 0000000000000246 R12: 00007ffd7c30dc58
[  103.110625] R13: 0000000000000000 R14: 000055fa23d8fb20 R15: 000000005db2db25
[  103.111092] Modules linked in: tun xt_TARPIT(OE) xt_NFLOG ipt_REJECT nf_reject_ipv4 xt_geoip(OE) xt_multiport iptable_filter xt_nat xt_comment xt_length iptable_nat nf_nat iptable_mangle xt_CT iptable_raw ip_tables ip6table_mangle ip6t_rt xt_conntrack nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 ip6table_filter ip6_tables ipv6 bcachefs lz4_compress xor zstd_decompress zstd_compress xxhash zlib_deflate raid6_pq lz4_decompress crct10dif_pclmul ghash_clmulni_intel aesni_intel aes_x86_64 glue_helper crypto_simd tcp_diag cryptd inet_diag virtio_balloon processor button sd_mod virtio_net net_failover crc32c_intel failover virtio_console virtio_scsi ata_generic
[  103.113659] CR2: fffffffffffffffd
[  103.113921] ---[ end trace b799c3c95a655abc ]---
[  103.114216] RIP: 0010:bch2_inode_find_by_inum_trans+0x88/0xb0 [bcachefs]
[  103.114553] Code: ff ff 48 89 c3 41 89 c5 76 12 48 8d 65 d8 44 89 e8 5b 41 5a 41 5c 41 5d 41 5e 5d c3 48 89 c7 41 bd fe ff ff ff e8 38 c4 fc ff <80> 78 02 08 48 89 d6 74 0d 48 89 de 4c 89 e7 e8 f4 c9 fc ff eb ca
[  103.115469] RSP: 0018:ffffafaa405c3420 EFLAGS: 00010246
[  103.115757] RAX: fffffffffffffffb RBX: ffffafaa405c3500 RCX: 0000000000000001
[  103.116145] RDX: 0000000000000000 RSI: 0000000000000001 RDI: ffff9ac97f1a0000
[  103.116488] RBP: ffffafaa405c3460 R08: 0000000000000000 R09: ffffffffc04a80ff
[  103.116831] R10: 0000000000000000 R11: ffffafaa405c3500 R12: ffffafaa405c3470
[  103.117235] R13: 00000000fffffffe R14: ffffafaa405c38d0 R15: ffffafaa405c3d38
[  103.117583] FS:  00007fba311dd900(0000) GS:ffff9ac67fcc0000(0000) knlGS:0000000000000000
[  103.118031] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  103.118392] CR2: fffffffffffffffd CR3: 00000001fbb1e001 CR4: 0000000000360ea0
[  103.118736] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  103.119168] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  114.511765] zram: Added device: zram0
```

After another, gracefull  restart of vm I have:
```
[   17.051839] bcachefs (dm-7): journal read done, 0 keys in 1 entries, seq 1534317
[   17.064712] bcachefs (dm-7): starting metadata mark and sweep
[   17.094804] bcachefs (dm-7): error validating btree node at btree 1 level 0/1
               pos 12101:0 node offset 372 bset u64s 1476: invalid checksum
[   17.094808] bcachefs (dm-7): error validating btree node at btree 1 level 0/1
               pos 12101:0 node offset 372 bset u64s 1476: KEY_U64s 0: 5688 bytes of metadata lost
[   17.094810] bcachefs (dm-7): emergency read only
[   17.094831] bcachefs (dm-7): Error in recovery: error in recovery (-5)
[   17.095671] bcachefs (dm-7): error invalidating buckets: -30
[   17.219666] bcachefs (dm-6): recovering from clean shutdown, journal seq 7194
[   17.249258] bcachefs (dm-6): mounted with opts: compression=zstd,discard
[   17.319870] bcachefs (dm-4): recovering from clean shutdown, journal seq 1987990
[   17.582205] bcachefs (dm-4): mounted with opts: compression=zstd,discard
```

fsck can't fix it (I run fsck three times):
```
# bcachefs fsck -y /dev/dm-7
recovering from clean shutdown, journal seq 1534325
journal read done, 0 keys in 1 entries, seq 1534325
starting mark and sweep
error validating btree node at btree 0 level 0/1
pos 11612:104 node offset 380 bset u64s 426: invalid checksum
error validating btree node at btree 0 level 0/1
pos 11612:104 node offset 380 bset u64s 426: KEY_U64s 0: 1392 bytes of metadata lost, fixing
error validating btree node at btree 1 level 0/1
pos 12101:0 node offset 372 bset u64s 1476: invalid checksum
error validating btree node at btree 1 level 0/1
pos 12101:0 node offset 372 bset u64s 1476: KEY_U64s 0: 5688 bytes of metadata lost, fixing
starting fsck
mounted with opts: compression=zstd,degraded,fsck,fix_errors
```

**Comment by koverstreet** [comment]
Eesh. So, for the errors fsck isn't fixing, that's definitely memory corruption (something scribbling zeros over bcachefs metadata)... a bunch of stuff has been lost, not much we can do about that.

For the initial bug, that was caused by deferefing an -EINTR ptr - the code wasn't checking for errors from bch2_btree_iter_peek_slot(). We shouldn't be getting -EINTR there with only one iterator allocated, but the code should be checking for errors anyways... so I pushed a patch for that.
